src/Components/Web.Extensions/src/InputFile/RemoteBrowserFileStream.cs

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Web.Extensions
+{
+    internal class RemoteBrowserFileStream : BrowserFileStream
+    {
+        private readonly IJSRuntime _jsRuntime;
+        private readonly ElementReference _inputFileElement;
+        private readonly int _maxChunkSize;
+        private readonly PipeReader _pipeReader;
+        private readonly CancellationTokenSource _fillBufferCts;
+
+        private bool _isReadingCompleted;
+        private bool _isDisposed;
+
+        public RemoteBrowserFileStream(IJSRuntime jsRuntime, ElementReference inputFileElement, int maxChunkSize, int maxBufferSize, BrowserFile file)
+            : base(file)
+        {
+            _jsRuntime = jsRuntime;
+            _inputFileElement = inputFileElement;
+            _maxChunkSize = maxChunkSize;
+
+            var pipe = new Pipe(new PipeOptions(pauseWriterThreshold: maxBufferSize, resumeWriterThreshold: maxBufferSize));
+            _pipeReader = pipe.Reader;
+            _fillBufferCts = new CancellationTokenSource();
+
+            _ = FillBuffer(pipe.Writer, _fillBufferCts.Token);
+        }
+
+        private async Task FillBuffer(PipeWriter writer, CancellationToken cancellationToken)
+        {
+            long offset = 0;
+
+            while (offset < File.Size)
+            {
+                var pipeBuffer = writer.GetMemory(_maxChunkSize);
+                var chunkSize = (int)Math.Min(_maxChunkSize, File.Size - offset);
+
+                try
+                {
+                    var base64 = await _jsRuntime.InvokeAsync<string>(
+                        InputFileInterop.ReadFileData,
+                        cancellationToken,
```

Comment:
We should use a different cancellation token for this that ensures that the client remains responsive over the file upload.

If I allow to upload 1GB+ files, it is likely that I'll either pass down CancellationToken.None or a cancellation token for something like 30 minutes or 1h.

When that's the case, It's potentially possible for a client to send you lots of data and force you to hold on to a large amount of memory for a long time.

I would add this as an option in the options type that I described above, have it default to the same/similar value we use for the circuit and use a linked cancellation token between this value and the cancellation token passed in.